---
title: "01.1_reprocess_GSE150338"
author: "yincy"
date: "2023-04-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# prepare download links
```{r}
library(fs)
library(magrittr)
run_table <- read.table("data/GSE150338_SraRunTable.txt", sep = ",", header = TRUE)


base_url <- "https://trace.ncbi.nlm.nih.gov/Traces/sra-reads-be/fastq?acc="
urls <- paste(base_url, run_table$Run, sep = "")
write.table(urls, "data/GSE150338_fastq_urls.txt", row.names = F, col.names = F, quote = F)
```


# get sample info
```{r}
library(GEOquery)
library(janitor)
library(stringr.plus)

sample_info <- getGEO("GSE150338")
GSE150338_sample_info <- sample_info[[2]] %>% phenoData() %>% pData %>% 
    as.data.frame() %>% clean_names %>% 
    # dplyr::filter(source_name_ch1 %in% c("PTS1", "PTS2", "PTS3")) %>% 
    dplyr::select(title, source_name_ch1, sex_ch1, characteristics_ch1_4, characteristics_ch1
)

GSE150338_sample_info <- GSE150338_sample_info %>% 
    dplyr::rename("source" = "source_name_ch1", 
                  "sex" = "sex_ch1", 
                  "age" = "characteristics_ch1_4", 
                  "strain" = "characteristics_ch1")

GSE150338_sample_info <- GSE150338_sample_info %>% 
    dplyr::mutate(age = str_extract_after(age, ":"), 
                  strain = str_extract_after(strain, ":"))
write.table(GSE150338_sample_info, "../data/kidney/kidney-map/GSE150338/sample_info.txt", 
            sep = "\t", row.names = F, quote = F)
```


# quality control

# alignment

# analysis
```{r}
library(edgeR)
```








