---
title: "04 neurons data"
author: "YinCY"
date: "2023-03-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# GSE175421
Drop-seq sensory nerve from L1-L6
```{r}
library(magrittr)
library(dplyr)
library(fs)
library(data.table)
library(stringr.plus)
library(janitor)
library(purrr)
library(tidyr)

cell_barcode_ids <- read.csv("../data/kidney/kidney-map/GSE175421/GSE175421_cell-barcode-ids.csv.gz") %>% 
    clean_names()
cell_barcode_ids %>% dim
cell_barcode_ids %>% head
cell_barcode_ids$library_id %>% table

# select DRG info
drg_cell_barcode_ids <- cell_barcode_ids %>% dplyr::filter(grepl("^DRG", library_id, ignore.case = FALSE))
colnames(drg_cell_barcode_ids)[[1]] <- "cell_ids"
drg_cell_barcode_ids <- drg_cell_barcode_ids %>% 
    mutate(cell_ids = paste(str_extract_before(cell_ids, "_"), library_id, sep = "_"))
rownames(drg_cell_barcode_ids) <- drg_cell_barcode_ids$cell_ids
drg_cell_barcode_ids$cell_type <- drg_cell_barcode_ids$main_cluster
drg_cell_barcode_ids$cell_type %>% table
drg_cell_barcode_ids$cell_type <- ifelse(drg_cell_barcode_ids$cell_type == "Vascular Endothelial Cells", "Spinal Endothelial", drg_cell_barcode_ids$cell_type)
drg_cell_barcode_ids$location <- "Spine"

drg_cell_barcode_list <- split(drg_cell_barcode_ids, f = drg_cell_barcode_ids$library_id)

files <- dir_ls("../data/kidney/kidney-map/GSE175421/", regexp = "*.txt.gz$")
samples <- str_extract_between(basename(files), pattern1 = "_", pattern2 = "\\.", 
                               which_pattern1 = "first", which_pattern2 = "first") %>% 
    gsub("rep", "", .)
```


```{r}
# drg counts table
drg_list <- list()
for(i in seq_along(files)){
    cur <- fread(input = files[[i]], header = TRUE) %>% as.data.frame()
    cur <- cur %>% tibble::column_to_rownames("GENE")
    colnames(cur) <- paste(colnames(cur), samples[[i]], sep = "_")
    drg_list[[samples[[i]]]] <- cur
}
drg_cell_barcode_list %>% names; drg_list %>% names


# filter cells
for( i in samples){
    cur_counts <- drg_list[[i]]
    cur_info <- drg_cell_barcode_list[[i]]
    cur_counts <- cur_counts %>% dplyr::select(cur_info$cell_ids)
    drg_list[[i]] <- cur_counts
}

# merge samples
ugene <- sapply(drg_list, rownames) %>% unlist(use.names = F) %>% unique()
ugene %>% str
ugene <- data.frame(symbol = ugene)

library(AnnotationHub)
ah <- AnnotationHub()
query(ah, c('EnsDb', "Mus musculus"))
ensdb <- ah[["AH104895"]]

ugene$ensembl <- mapIds(ensdb, keys = ugene$symbol, 
                        keytype = "SYMBOL", 
                        column = "GENEID")
ugene <- ugene %>% dplyr::filter(!is.na(ensembl))
ugene$ensembl %>% duplicated() %>% table

drg_list <- lapply(drg_list, function(x){tibble::rownames_to_column(x, "symbol")}) %>% 
    sapply(., function(x){
    x = left_join(ugene, x)
    rownames(x) <- x$ensembl
    x <- x[, -c(1, 2)]
})

drg_list %>% sapply(dim)
sapply(drg_list, rownames) %>% apply(1, function(x){length(unique(x))}) %>% table
# all rownames are the same and in the same order

drg_counts_table <- map_dfc(drg_list, dplyr::bind_cols)
drg_counts_table[1:10, 1:5]
drg_counts_table %>% dim

# check the exist of NA
drg_counts_table %>% apply(2, is.na) %>% sum

# replace NA with 0
drg_counts_table <- drg_counts_table %>% apply(2, function(x){
    replace_na(x, 0)
})
```


```{r}
# converted to SCE
gene_df <- data.frame(ensembl = rownames(drg_counts_table), 
                      symbol = mapIds(ensdb, 
                                      keys = rownames(drg_counts_table), 
                                      keytype = "GENEID", 
                                      column = "SYMBOL"))
library(SingleCellExperiment)
library(Matrix)
drg_sce <- SingleCellExperiment(assays = list(counts = as(as.matrix(drg_counts_table), "CsparseMatrix")), 
                                rowData = gene_df, 
                                colData = drg_cell_barcode_ids)

is_mito <- grepl("^mt-", rowData(drg_sce)$symbol)
is_mito %>% table
rowData(drg_sce)$symbol[is_mito] %>% sort
colData(drg_sce)$mito_percent <- round((colSums2(assay(drg_sce[is_mito, ])) / colSums2(assay(drg_sce))) * 100, 4)
colData(drg_sce)
```


```{r}
# remove Fibroblasts, Macrophages, Mural Cells, Satellite Glia, T Cells, Sympathetic Neurons
drg_sce %>% colData %>% .$cell_type %>% table
cells_to_remove <- c("Fibroblasts", "Macrophages", "Mural Cells", "T Cells", "Sympathetic Neurons")
drg_sce <- drg_sce[, !(drg_sce$main_cluster %in% cells_to_remove)]

# normalization
library(scuttle)
library(BiocParallel)
set.seed(101)
lib_size <- pooledSizeFactors(counts(drg_sce), clusters = drg_sce$cell_type)
lib_size %>% quantile(probs = seq(0, 1, 0.1))
sizeFactors(drg_sce) <- lib_size
drg_sce <- logNormCounts(drg_sce)

# model gene variance
library(scran)
set.seed(102)
variance <- modelGeneVar(drg_sce, block = drg_sce$library_id)
variance[order(variance$bio, decreasing = T), ]
hvgs <- getTopHVGs(variance, fdr.threshold = 0.01)
hvgs %>% str

# dimensional reduction
library(scater)
library(PCAtools)
set.seed(103)
npcs <- calculatePCA(x = logcounts(drg_sce), subset_row = hvgs)
elbow <- findElbowPoint(attr(npcs, "percentVar"))
reducedDim(drg_sce, "PCA") <- npcs[, 1:10]
plot(attr(npcs, "percentVar"))

set.seed(104)
drg_sce <- runTSNE(drg_sce, dimred = "PCA", perplexity = 20)

# clusters
library(bluster)
set.seed(105)
clusters <- clusterRows(reducedDim(drg_sce, "PCA"), 
                        BLUSPARAM = NNGraphParam(k = 30))
clusters %>% table
colData(drg_sce)$clusters <- clusters

colData(drg_sce)$cell_type <- sapply(colData(drg_sce)$clusters, 
                                        FUN = function(x){
                                            switch(x, 
                                                   "1" = "Sensory Neurons", 
                                                   "2" = "Satellite Glia B", 
                                                   "3" = "Sensory Neurons", 
                                                   "4" = "Satellite Glia A", 
                                                   "5" = "Satellite Glia A", 
                                                   "6" = "Schwann Cells", 
                                                   "7" = "Schwann Cells", 
                                                   "8" = "Spinal Endothelial", 
                                                   "9" = "Sensory Neurons", 
                                                   "10" = "Spinal Endothelial")
                                        })
colData(drg_sce)$cell_type %>% table
source("functions/idconv.R")
plotTSNE(drg_sce, 
         colour_by = "cell_type",
         # colour_by = idconv("Calca"),
         point_size = 0.5, 
         text_by = "cell_type")
saveRDS(drg_sce, "../data/kidney/kidney-map/GSE175421/drg_sce.rds")
```



# GSE78845
Fluidigm C1 sequencer
T1-T12 sympathetic ganglia

# reanalysis GSE78845
```{r}
library(fs)
library(stringr)

files <- dir_ls("../data/kidney/kidney-map/GSE78845/fastq/")
files_to <- paste0("../data/kidney/kidney-map/GSE78845/fastq/", str_replace(basename(files), "^fastq\\?acc=", ""), ".fastq.gz")
file_move(path = files, new_path = files_to)
```

```{r}
knitr::include_graphics("../BioR/bulk-seq/fastqcr/fastqcr.png")
```


# quality control
```{r}
library(fastqcr)
library(fs)
library(magrittr)
library(ggplot2)

fastqc(fq.dir = "../data/kidney/kidney-map/GSE78845/fastq/", 
       qc.dir = "../data/kidney/kidney-map/GSE78845/QC/",
       threads = 8, 
       fastqc.path = "/usr/bin/fastqc")

qc_report(qc.path = "../data/kidney/kidney-map/GSE78845/QC/", 
          result.file = "../data/kidney/kidney-map/GSE78845/")

qc_collection <- qc_read_collection(files = dir_ls("../data/kidney/kidney-map/GSE78845/QC/", regexp = "zip$"), 
                                    sample_names = basename(dir_ls("../data/kidney/kidney-map/GSE78845/QC/", regexp = "zip$")),
                                    modules = "all", 
                                    verbose = F)
saveRDS(qc_collection, "../data/kidney/kidney-map/GSE78845/QC-out/qc_collection.rds")

base_quality <- qc_plot_collection(qc = qc_collection, 
                   modules = "Per base sequence quality")

base_quality + theme(legend.position = "none")

# The first 6 base of each read represent the random Unique Molecular Identifier 
# user for molecule counting. After following three or more Gs, stemming the 
# template switching at the mRNA's 5' ending during first strand cDNS sythesis.
```


# alignment and feature counts
```{r}
library(Rsubread)
library(stringr)
library(fs)

# build index
# buildindex(basename = "../../Documents/Biology/reference-genome/mouse/ensembl/ensembl_GRCm39_reference_index_subread", 
#           reference = "../../Documents/Biology/reference-genome/mouse/ensembl/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz")

files <- dir_ls("../data/kidney/kidney-map/GSE78845/fastq/")
align(index = "../../Documents/Biology/reference-genome/mouse/ensembl/ensembl_GRCm39_reference_index_subread", 
      readfile1 = files, 
      output_file = paste0("../data/kidney/kidney-map/GSE78845/bam/", str_replace(basename(files), "fastq.gz", "bam")), 
      nthreads = 8, 
      nTrim5 = 11)

feature_counts <- featureCounts(files = paste0("../data/kidney/kidney-map/GSE78845/bam/", str_replace(basename(files), "fastq.gz", "bam")), 
                                # annot.inbuilt = "mm39", 
                                annot.ext = "../../Documents/Biology/reference-genome/mouse/ensembl/Mus_musculus.GRCm39.109.gtf.gz", 
                                isGTFAnnotationFile = T)

saveRDS(feature_counts, "../data/kidney/kidney-map/GSE78845/feature_counts.rds")
```


# create SCE object
```{r}
library(magrittr)
library(dplyr)
library(janitor)

run_info <- read.table("data/GSE78845_SraRunTable.txt", sep = ",", header = TRUE) %>% 
    clean_names()
selected_vars <- c("run", "cell_type", "celltype_identifier", 
                   "sex", "source_name", "age")
run_info <- run_info %>% select(selected_vars) %>% 
    rename(cell_ids = "run")
run_info$cell_type <- ifelse(run_info$cell_type == "sympathetic neuronal cells", "Sympathetic Neurons", run_info$cell_type)
run_info$location <- "Spine"
rownames(run_info) <- run_info$cell_ids

feature_counts <- readRDS("../data/kidney/kidney-map/GSE78845/feature_counts.rds")
gene_df <- data.frame(ensembl = rownames(feature_counts$counts))

library(AnnotationHub)
ah <- AnnotationHub()
query(ah, c("EnsDb", "Mus musculus"))
ensdb <- ah[["AH104895"]]

gene_df$symbol <- mapIds(ensdb, 
                         keys = gene_df$ensembl, 
                         keytype = "GENEID", 
                         column = "SYMBOL")
rownames(gene_df) <- gene_df$ensembl

mtx <- feature_counts$counts
mtx[1:10, 1:5]
colnames(mtx) <- gsub(".bam", "", colnames(mtx))


library(SingleCellExperiment)
library(Matrix)

sym_sce <- SingleCellExperiment(assays = list(counts = as(mtx, "CsparseMatrix")), 
                                rowData = gene_df, 
                                colData = run_info)

sym_sce
```


```{r}
# quality control
library(scuttle)
library(scran)
library(scater)
library(magrittr)

set.seed(101)
feature_stats <- perFeatureQCMetrics(sym_sce)
filter_index <- feature_stats$mean >= 0.2 & feature_stats$detected >= 5
sym_sce <- sym_sce[filter_index, ]

is.mito <- grepl("^mt-", rowData(sym_sce)$symbol, ignore.case = F)
is.mito %>% table
colData(sym_sce)$mito_percent <- round(colSums(counts(sym_sce)[is.mito, ]) / colSums(counts(sym_sce)), digits = 5) * 100
sym_sce %>% colData

set.seed(102)
cell_stats <- perCellQCMetrics(sym_sce, subset = list(mito = is.mito))
qc <- perCellQCFilters(cell_stats, sub.fields = list('subsets_mito_percent'))
qc %>% as.data.frame() %>% colSums()

sym_sce <- sym_sce[, !qc$discard]

set.seed(103)
clusters <- quickCluster(sym_sce, min.size = 20)
clusters %>% str
colData(sym_sce)$qclusters <- factor(clusters)

table(colData(sym_sce)$qclusters, colData(sym_sce)$celltype_identifier)

set.seed(104)
sym_sce <- computePooledFactors(sym_sce, clusters = sym_sce$qclusters)
sym_sce <- logNormCounts(sym_sce)

# model gene variance
set.seed(105)
dec_var <- modelGeneVar(sym_sce, block = sym_sce$sex)
dec_var[order(dec_var$bio, decreasing = T), ]

hvgs <- getTopHVGs(dec_var, fdr.threshold = 0.01)
hvgs %>% str

# dimensional reduction
set.seed(106)
npcs <- calculatePCA(logcounts(sym_sce), ncomponents = 20, subset_row = hvgs)
library(PCAtools)
elbow <- findElbowPoint(attr(npcs, "percentVar"))
reducedDim(sym_sce, "PCA") <- npcs[, 1:elbow]

set.seed(107)
sym_sce <- runTSNE(sym_sce, dimred = "PCA", perplexity = 10)

source("R/idconv.R")
plotTSNE(sym_sce, 
         # colour_by = "qclusters", 
         colour_by = idconv("Dbh"))

sym_sce %>% colData()
saveRDS(sym_sce, "../data/kidney/kidney-map/GSE78845/sym_sce.rds")
```







