Single Cell RNASeq Data Analysis 
Principle component analysis and identification of variably expressed genes were carried out using the R packages Seurat (v 2.3.4) (Butler et al., 2018), ggplot2 (v. 3.0), Matrix (v.1.2-14) and dplyr (v 0.7.5) in R Studio. Briefly, for the primary analysis, the matrices from the 12 individual samples were first merged into a single metadata annotated r-object of 40,712 cells using the Read10X, CreateSeuratObject AddMetaData and MergeSeurat functions. The raw dataset was then filtered to remove genes expressed in less than three cells and the FilterCells function run with range filters for genes/cell (1,000-4,000) and mRNA transcripts/cell (UMIs; 1,000-16,000), and *maximum percent mitochondrial genes/cell (35%)*; note renal epithelial cell types have high metabolic rates and expected higher mitochondrial gene detection than other cell types (Figures S1Dâ€“S1F). The resulting primary analysis dataset contained 19,125 genes across 31,265 cells. The NormalizeData function scaled and log transformed that dataset, from which *2464 variably expressed genes* were identified with FindVariableGenes, followed by performing principle component analysis first on the variable genes with RunPCA, then extending to all dataset genes with ProjectPCA. *The 30 useful principle components determined by the visualization tools PCElbowPlot and JackStrawPlot were used in graph-based clustering using FindClusters at resolution 1.0*. FindAllMarkers generated the list of genes differentially expressed in each cluster compared to all other cells (Table S1) based on the Wilcoxon rank-sum test and limiting the analysis with a cutoff for minimum log FC difference (0.25) and minimum cells with expression (0.25). The secondary Seurat analyses of the nephron and ureteric epithelium (UE) profiles employed the SubsetData function to create new r-objects from cohorts of primary analysis clusters identified as nephron (4,6,8,10,12,13,18,25,26,27) or UE (5,7,20,23,17). Re-processing these with Seurat tools, as described above, identified contaminating non-type cells (vascular and immune) or unclassifiable profiles (stressed or mitotic) that were removed to yield the nephron dataset with 8,854 cells and the UE dataset with 4,349 cells. Visualizations using the DotPlot function and ggplot2 package graphically represent per cluster percentage expression (pct. exp.) by dot diameter and average nonzero expression in log2 scale (avg. exp.) by dot color; no dot shown if pct. exp. < 0.05.