---
title: "mouse kidney"
author: "YinCY"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# data
1. GSE107585: whole mice kidney single cell sequencing data, 2018 science  
2. GSE129798: whole mice kidney single cell sequencing data, with regional information, 2019 development cell  
3. GSE160048: mice and human glomerular sequencing data  
    candidate: GSE174102, GSE127235, GSE201932
4. GSE175421: mice sympathetic and sensory nerve sequencing data  
5. PRJCA003940: embryonic and monocytes derived macrophage bulk sequencing data, for reference  
6. GSE150338: microdissected mouse tubules sequencing data, for reference  

remove some very few cells from each dataset.


# prepare mouse tubula sequencing data
```{r}
tubule_region <- data.frame(
    short_name = c("PTS1", "PTS2", "PTS3", "DTL1", "DTL2", "DTL3", "ATL", 
                   "MTAL", "CTAL", "DCT", "CNT", "CCD", "OMCD", "IMCD"), 
    log_name = c("initial segment of the proximal convoluted tubule",
                 "proximal straight tubule in cortical medullary lays", 
                 "last segment of the proximal straight tubule in outer stripe of outer medulla", 
                 "short descending limb of the loop of Henle", 
                 "long descending limb of the loop of Henle in the outer medulla", 
                 "long descending limb of the loop of Henle in the inner medulla", 
                 "thin ascending limb of the loop of Henle", 
                 "medullary thick ascending limb of the loop of Henle", 
                 "cortical thick ascending limb of the loop of Henle", 
                 "distal convoluted tubule", 
                 "connecting tubule", 
                 "cortical collecting duct", 
                 "outer medullary collecting duct", 
                 "inner medullary collecting duct")
)

tubule_region
```

```{r}
library(dplyr)
library(stringr)

mouse_tubule <- read.csv("../../git/data/kidney/kidney-map/mouse-tubule/GSE150338_Mouse_TPM_Replicates.csv", 
                         sep = "\t")
mouse_tubule %>% dim
mouse_tubule %>% tail
mouse_tubule %>% head
mouse_tubule %>% colnames()
mouse_tubule$X %>% table(useNA = "ifany")
mouse_tubule

# discate glomerila 
mouse_tubule <- mouse_tubule %>% select(!matches("_GL_|X"))
rownames(mouse_tubule) <- mouse_tubule$gene_id
rowdata <- mouse_tubule[, 1:2]
coldata <- data.frame(sample = colnames(mouse_tubule[, -c(1:2)]), 
                      region = gsub("m[0-9]{0,}_|_[a-z]{0,}", replacement = "", colnames(mouse_tubule[, -c(1:2)])), 
                      row.names = colnames(mouse_tubule[, -c(1:2)]))

ref_mouse_tubule_sce <- SingleCellExperiment(assays = list(tpm = mouse_tubule[, -c(1:2)]), 
                                             rowData = rowdata, 
                                             colData = coldata)
rowData(ref_mouse_tubule_sce)

saveRDS(ref_mouse_tubule_sce, 
        file = "../../git/data/kidney/kidney-map/mouse-tubule/ref_mouse_tubule_sce.rds")
```


# prepare dev cell data
```{r}
library(DropletUtils)
library(scran)
library(scuttle)
library(scater)
library(magrittr)

dev_cell_sce <- read10xCounts("../data/kidney/kidney-map/dev-cell/Mouse_Adult_DGE/")
dev_cell_sce %>% colData()
dev_cell_sce %>% rowData() %>% .$ID %>% duplicated() %>% table

# use the newest annotation data
library(AnnotationHub)
ah <- AnnotationHub()
query(ah, c("EnsDb", "Mus Musculus"))
ens107 <- ah[["AH104895"]]
ens107 %>% columns

rowData(dev_cell_sce)$ensembl <- mapIds(ens107, 
                                        keys = rowData(dev_cell_sce)$ID, 
                                        keytype = "SYMBOL", 
                                        column = "GENEID")

rowData(dev_cell_sce)$ensembl %>% is.na() %>% table()
dev_cell_sce <- dev_cell_sce[!is.na(rowData(dev_cell_sce)$ensembl), ]
rownames(dev_cell_sce) <- rowData(dev_cell_sce)$ensembl
rowData(dev_cell_sce)

colData(dev_cell_sce)$region <- substr(colData(dev_cell_sce)$Barcode, 1, 2)
colData(dev_cell_sce)$batch <- substr(colData(dev_cell_sce)$Barcode, 1, 4)
colData(dev_cell_sce)$gender <- substr(colData(dev_cell_sce)$Barcode, 4, 4)
colData(dev_cell_sce)

# retain only the male data
dev_cell_sce <- dev_cell_sce[, dev_cell_sce$gender == "M"]

table(colData(dev_cell_sce)$gender);table(colData(dev_cell_sce)$batch);table(colData(dev_cell_sce)$region)
```

```{r}
# quality control
is_mito <- grepl("^mt-", rowData(dev_cell_sce)$ID, ignore.case = F)
is_mito %>% table
rowData(dev_cell_sce)$ID[is_mito] %>% sort

qclusters <- quickCluster(dev_cell_sce, 
                         min.size = 1000, 
                         method = "igraph", 
                         use.ranks = T, 
                         d = 50)
qclusters %>% table
colData(dev_cell_sce)$quick_cluster <- qclusters

library(tibble)
colData(dev_cell_sce) %>% 
    as_tibble() %>% 
    ggplot(aes(region, quick_cluster)) +
    geom_jitter(size = 0.1)

colData(dev_cell_sce)$mito_percent <- round((colSums2(assay(dev_cell_sce[is_mito, ])) / colSums2(assay(dev_cell_sce))) * 100, 4)

colData(dev_cell_sce) %>% 
    as_tibble() %>% 
    ggplot(aes(quick_cluster, mito_percent)) +
    geom_jitter(size = 0.1)

# source("R/splitCols.R")
# sces <- splitCols(dev_cell_sce, f = qclusters)
# 
# qc_list <- vector(mode = "list")
# is_mito_list <- split(is_mito, qclusters)
# for(name in names(sces)){
#     qc_list[[name]] <- perCellQCMetrics(sces[[name]], 
#                                      subsets = list(mito = is_mito_list[[name]]))
# }
# 
# qc_filter_list <- vector(mode = "list", length = length(qc_list))
# names(qc_filter_list) <- names(qc_list)
# for(name in names(qc_list)){
#     qc_filter_list[[name]] <- perCellQCFilters(qc_list[[name]], sub.fields = TRUE)
# }
# qc_filter_list %>% sapply(FUN = function(x){
#     x[["discard"]] %>% table
# })

dev_cell_sce <- addPerCellQCMetrics(dev_cell_sce, subsets = list(mito = is_mito))
dev_cell_sce %>% colData
qc <- perCellQCMetrics(dev_cell_sce, subsets = list(mito = is_mito))
qc_filter <- perCellQCFilters(qc, sub.fields = TRUE, log = TRUE)
qc_filter %>% as.data.frame() %>% colSums()

library(patchwork)
dev_cell_sce[, qc_filter$discard] %>% 
    colData() %>% 
    as_tibble() %>% 
    ggplot(aes(quick_cluster, mito_percent)) +
    geom_jitter(width = 0.2, size = 0.1) +
dev_cell_sce[, !qc_filter$discard] %>% 
    colData() %>% 
    as_tibble() %>% 
    ggplot(aes(quick_cluster, mito_percent)) +
    geom_jitter(width = 0.2, size = 0.1) +
dev_cell_sce %>% 
    colData() %>% 
    as_tibble() %>% 
    ggplot(aes(detected, mito_percent)) +
    geom_jitter(width = 0.1, size = 0.1) +
    geom_vline(xintercept = 1000) +
    geom_hline(yintercept = 50) +
dev_cell_sce %>% 
    colData %>% 
    as_tibble %>% 
    ggplot(aes(total, mito_percent)) +
    geom_jitter(width = 0.1, size = 0.1) +
    geom_vline(xintercept = 5000) +
    geom_hline(yintercept = 50)

dev_cell_sce %>% 
    colData() %>% 
    as_tibble() %>% 
    ggplot(aes(detected, mito_percent)) +
    geom_jitter(aes(colour = qc_filter$low_n_features), width = 0.1, size = 0.1)

# filter based on number of features
dev_cell_sce <- dev_cell_sce[, !qc_filter$low_n_features]

saveRDS(dev_cell_sce, 
        "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```

```{r}
# normalization
dec_sf_dev_cell <- calculateSumFactors(dev_cell_sce, clusters = colData(dev_cell_sce)$quick_cluster)
sizeFactors(dev_cell_sce) <- dec_sf_dev_cell
colData(dev_cell_sce)
dev_cell_sce <- computeLibraryFactors(dev_cell_sce)

dev_cell_sce <- logNormCounts(dev_cell_sce)
dev_cell_sce
saveRDS(dev_cell_sce, "../data/kidney/kidney-map/dev-cell/dev_cell_sce_0114.rds")
```

```{r}
# feature selection
library(magrittr)
library(stringr)
library(scuttle)
library(scran)
library(scater)
library(BiocParallel)

colData(dev_cell_sce)
dec_dev_cell <- modelGeneVar(dev_cell_sce, block = colData(dev_cell_sce)$batch)
dec_dev_cell[order(dec_dev_cell$bio, decreasing = T), ]

dec_pos_dev_cell <- modelGeneVarByPoisson(dev_cell_sce, size.factors = dev_cell_sce$sizeFactor, block = colData(dev_cell_sce)$batch)
dec_pos_dev_cell[order(dec_pos_dev_cell$bio, decreasing = T), ]

# dec_cv2_dev_cell <- modelGeneCV2(x = dev_cell_sce, 
#                                  size.factor = sizeFactors(dev_cell_sce))
# dec_cv2_dev_cell[order(dec_cv2_dev_cell$FDR, decreasing = F), ]
```

```{r}
par(mfrow = c(1, 3))
for(i in 1:ncol(dec_dev_cell$per.block)){
    plot(metadata(dec_dev_cell$per.block[[i]])$mean, 
         metadata(dec_dev_cell$per.block[[i]])$var, 
         pch = 19, cex = 0.5, 
         main = paste0("model_", names(dec_dev_cell$per.block)[[i]]))
    curve(metadata(dec_dev_cell$per.block[[i]])$trend(x), col = "red", lwd = 2, add = TRUE)
}
```


```{r}
par(mfrow = c(1, 3))
for(i in 1:ncol(dec_pos_dev_cell$per.block)){
    plot(metadata(dec_pos_dev_cell$per.block[[i]])$mean, 
         metadata(dec_pos_dev_cell$per.block[[i]])$var, 
         pch = 19, cex = 0.5, 
         main = paste0("Pois_", names(dec_pos_dev_cell$per.block)[[i]]))
    curve(metadata(dec_pos_dev_cell$per.block[[i]])$trend(x), col = "red", lwd = 2, add = TRUE)
}
```

```{r}
# par(mfrow = c(2, 3))
# for(i in 1:ncol(dec_cv2_dev_cell$per.block)){
#     plot(metadata(dec_cv2_dev_cell$per.block[[i]])$mean, 
#          metadata(dec_cv2_dev_cell$per.block[[i]])$var, 
#          pch = 19, lwd = 2, 
#          main = paste0("CV2_", names(dec_cv2_dev_cell$per.block)[[i]]))
#     curve(metadata(dec_cv2_dev_cell$per.block[[i]])$trend(x), col = "red", lwd = 2, add = TRUE)
# }
```



```{r}
dev_cell_hvgs <- getTopHVGs(stats = dec_dev_cell, 
                            var.field = "bio",
                            var.threshold = 0, 
                            fdr.threshold = 0.01)
dev_cell_hvgs %>% str

# choosing the number of PCs
dev_cell_sce <- runPCA(dev_cell_sce, subset_row = dev_cell_hvgs, ncomponents = 50)
percentvar <- reducedDim(dev_cell_sce, "PCA") %>% attr(which = "percentVar")

library(PCAtools)
elbow <- findElbowPoint(percentvar)

df <- data.frame(PCs = 1:dim(reducedDim(dev_cell_sce, "PCA"))[2], 
                 percentvar = percentvar)

df %>% 
    ggplot(aes(PCs, percentvar)) + 
    geom_line() + 
    geom_point() +
    geom_vline(xintercept = elbow) +
    scale_y_continuous(trans = "log2")

npcs <- denoisePCA(dev_cell_sce, technical = dec_pos_dev_cell, subset.row = dev_cell_hvgs) %>% 
    reducedDim(type = "PCA") %>% ncol

pcs <- reducedDim(dev_cell_sce, "PCA")

library(bluster)
choise <- getClusteredPCs(pcs = as.matrix(pcs), 
                          BLUSPARAM = SNNGraphParam(type = "jaccard", k = 10, BPPARAM = MulticoreParam(workers = 12)))
choise %>% metadata()

dev_cell_sce <- runPCA(x = dev_cell_sce, 
                       subset_row = dev_cell_hvgs, 
                       ncomponents = 9, 
                       BPPARAM = MulticoreParam(workers = 12))

# determine perplexity
per <- seq(10, 80, 10) %>% as.character()
sces <- vector(mode = "list")
for(i in per){
    sces[[i]] <- runTSNE(dev_cell_sce, 
                         dimred = "PCA", 
                         perplexity = as.integer(i), 
                         BPPARAM = MulticoreParam(workers = 12))
}

p <- lapply(sces, FUN = plotTSNE, 
            colour_by = "quick_cluster", 
            text_by = "quick_cluster")


# using the preplexity of 10
set.seed(110)
dev_cell_sce <- runTSNE(x = dev_cell_sce, 
                        dimred = "PCA", 
                        perplexity = 10, 
                        BPPARAM = MulticoreParam(workers = 12))
plotTSNE(dev_cell_sce, 
         colour_by = "quick_cluster", 
         point_size = 0.2, 
         text_by = "quick_cluster")
saveRDS(dev_cell_sce, "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```


```{r}
# dev_cell_sce <- readRDS("../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
# graph based clustering
library(scran)
library(igraph)
library(scater)
library(BiocParallel)

grep("^cluster_", ls("package:igraph"), value = T)

graph <- buildSNNGraph(dev_cell_sce, k = 10, use.dimred = "PCA")
clusters <- cluster_louvain(graph)$membership
clusters %>% table

colLabels(dev_cell_sce) <- factor(clusters)
colData(dev_cell_sce)
```

```{r}
# highlight each cluster and sub-cluster
for( n in levels(colLabels(dev_cell_sce))){
    my_color <- rep("grey", length(colLabels(dev_cell_sce)))
    names(my_color) <- as.character(colLabels(dev_cell_sce))
    my_color[colLabels(dev_cell_sce) == n] <- "red"
    
    plotReducedDim(object = dev_cell_sce, 
                   dimred = "TSNE", 
                   colour_by = "label", 
                   text_by = "label", 
                   point_size = 0.2) +
        scale_color_manual(values = my_color) +
        guides(colour = guide_legend(override.aes = list(size = 2)))
    
    ggsave(paste0("dev_image/the-", n, "th-cluster.pdf"))
}
```


```{r}
# doublet detection
library(scDblFinder)
dbs <- findDoubletClusters(dev_cell_sce)
outs <- rownames(dbs)[isOutlier(dbs$median.de, type = "lower", log = TRUE)]

# cluster 2 is doublets
# remove cluster 2
dev_cell_sce <- dev_cell_sce[, colLabels(dev_cell_sce) != 2]
colLabels(dev_cell_sce) <- factor(colLabels(dev_cell_sce), levels = c(1, 3:19))
dev_cell_sce %>% colData()
```


```{r}
# marker gene detection
markers <- scoreMarkers(x = dev_cell_sce, 
                        groups = colLabels(dev_cell_sce))

source("R/idconv.R")
markers[["15"]][order(markers[["15"]]$min.AUC, decreasing = T), ] %>% 
    head(50) %>% 
    as.data.frame() %>% 
    rownames() %>% 
    idconv(from_type = "GENEID", to_type = "SYMBOL")

fmarkers <- findMarkers(dev_cell_sce, 
                        groups = colLabels(dev_cell_sce), 
                        test.type = "wilcox", 
                        pval.type = "all", 
                        direction = "up", 
                        min.prop = 0.25)
fmarkers[["5"]] %>% 
    as.data.frame() %>% 
    dplyr::arrange(FDR) %>% 
    head(50) %>% 
    rownames() %>% 
    idconv(from_type = "GENEID", to_type = "SYMBOL")

plotReducedDim(dev_cell_sce, 
               colour_by = idconv("Mcam"),
               # colour_by = "mito_percent",
               text_by = "label", 
               dimred = "TSNE", 
               point_size = 0.2)

idconv("Mcam")
ggcells(dev_cell_sce, 
        aes(colLabels(dev_cell_sce), ENSMUSG00000029231)) +
    geom_violin() +
    geom_jitter(size = 0.1)
saveRDS(dev_cell_sce, "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```

# sub-clusters
```{r}
# 5
c5 <- dev_cell_sce[, colLabels(dev_cell_sce) == 5]
hvgs <- c("H2-Ab1" = "ENSMUSG00000073421",  "H2-Eb1" = "ENSMUSG00000060586", "Csf1r" = "ENSMUSG00000024621", 
          "Fcgr2b" = "ENSMUSG00000026656", "Fcgr3" = "ENSMUSG00000059498", "Cx3cr1" = "ENSMUSG00000052336")
c5 <- runPCA(c5, exprs_values = "logcounts", 
             ncomponents = 2, 
             subset_row = hvgs)
c5 <- runTSNE(c5, perplexity = 30, dimred = "PCA")

g_c5 <- buildSNNGraph(c5, k = 100, use.dimred = "TSNE")
clusters_c5 <- cluster_louvain(g_c5)$membership
clusters_c5 %>% table
colLabels(c5) <- ifelse(clusters_c5 == 3, 2, clusters_c5)
colLabels(c5) <- paste("5.", colLabels(c5), sep = "")
labels <- as.character(colLabels(dev_cell_sce))
labels[labels == "5"] <- colLabels(c5)
colLabels(dev_cell_sce) <- labels
plotReducedDim(c5, dimred = "TSNE", colour_by = idconv("H2-Ab1"), text_by = "label")


# 16
hvgs <- c("Cspg4" = "ENSMUSG00000032911", 
          "Cnn1" = "ENSMUSG00000001349", 
          "Pdgfrb" = "ENSMUSG00000024620", 
          "Des" = "ENSMUSG00000026208", 
          "Mcam" = "ENSMUSG00000032135")
c16 <- dev_cell_sce[, colLabels(dev_cell_sce) == 16]
c16 <- runPCA(c16, subset_row = hvgs, ncomponents = 2, exprs_values = "logcounts")
c16 <- runTSNE(c16, perplexity = 30, use.dimred = "PCA")
g_c16 <- buildSNNGraph(c16, k = 100, use.dimred = "TSNE")
clusters_c16 <- cluster_louvain(g_c16)$membership
clusters_c16 %>% table
colLabels(c16) <- clusters_c16
plotReducedDim(c16, colour_by = idconv("Cnn1"), text_by = "label", dimred = "TSNE")

labels <- colLabels(dev_cell_sce)
labels[labels == 16] <- paste("16.", clusters_c16, sep = "")
labels %>% table
colLabels(dev_cell_sce) <- factor(labels, levels = c(1, 3, 4, 5.1, 5.2, 6:15, 16.1, 16.2, 17:19))
colLabels(dev_cell_sce)
saveRDS(dev_cell_sce, "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```



```{r}
# cell type annotation base on marker genes
markers <- list(
    endothelia = c("Cdh5", "Pecam1", "Kdr", "Nrp1"), 
    macrophage = c("C1qa", "C1qb", "C1qc", "Cx3cr1", "Lyz2"), 
    T_cell = c("Thy1", "Cd3g", "Cd3d"), 
    B_cell = c("Cd79a", "Cd79b"),
    NK = c("Tyrobp"), 
    DC = c("H2-Eb1", "H2-Ab1"),
    Monocytes = c("Csf1r"),
    VSMC = c("Cnn1"), 
    pericyte = c("Cspg4"), 
    Fibroblast = c("Dcn"), 
    PTS1 = c("Slc5a2", "Slc5a12", "Slc7a7", "Spp2", "Slc6a19", "Slc22a29"),
    PTS2 = c("Fxyd2", "Hrsp12"), 
    PTS3 = c("Atp11a", "Slc13a3", "Slc27a2"),
    CD_IC = c("Atp6v1g3", "Slc4a1", "Aqp6", "Slc26a4", "Hmx2"),
    CD_PC = c("Hsd11b2", "Aqp4", "Aqp2"),
    DCT = c("Pvalb", "Slc12a3"),
    LOH = c("Clcnka", "Slc12a1", "Umod", "Sptssb")
)

cell_annotation <- c(
    "1" = "Endothelia",
    "3" = "PTS1",
    "4" = "CD_IC",
    "5.1" = "Dendritic cell",
    "5.2" = "Monocyte",
    "6" = "PTS3", 
    "7" = "Macrophage",
    "8" = "DCT",
    "9" = "Endothelia",
    "10" = "PTS2", 
    "11" = "CD_PC", 
    "12" = "LOH", 
    "13" = "B cell", 
    "14" = "T cell",
    "15" = "Fibroblast", 
    "16.1" = "VSMC",
    "16.2" = "Pericyte",
    "17" = "Endothelia",
    "18" = "Endothelia",
    "19" = "LOH"
)
cell_annotation
colData(dev_cell_sce)$cell_type <- cell_annotation[colLabels(dev_cell_sce)]
dev_cell_sce %>% colData()
saveRDS(dev_cell_sce, "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```


## annotate region information
```{r}
library(SingleR)
library(magrittr)

dev_cell_sce <- readRDS("../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
ref_mouse_tubule_sce <- readRDS("../data/kidney/kidney-map/mouse-tubule/ref_mouse_tubule_sce.rds")
colData(ref_mouse_tubule_sce)

cell_to_annotate <- c("PTS1", "PTS2", "PTS3", "LOH", "CD_PC", "CD_IC", "DCT")
subs_dev_cell_tubule <- dev_cell_sce[, dev_cell_sce$cell_type %in% cell_to_annotate]

tubule_pred <- SingleR(test = subs_dev_cell_tubule, 
                       ref = ref_mouse_tubule_sce, 
                       labels = ref_mouse_tubule_sce$region, 
                       assay.type.test = "logcounts", 
                       assay.type.ref = "tpm", 
                       de.method = "t", 
                       de.n = 15)

ref_mouse_tubule_sce %>% colData()
tubule_pred$labels %>% table
```

```{r}
# regional information
library(tibble)
tubule_region_info <- tibble(
    short_name = c("ATL", "CCD", "CNT", 
                   "CTAL", "DCT", "DTL1", 
                   "DTL2", "DTL3", "IMCD", 
                   "MTAL", "OMCD", "PTS1", 
                   "PTS2", "PTS3"), 
    long_name = c("thin ascending limb of the loop of Henle", 
                  "cortical collecting duct", 
                  "connecting tubule", 
                  "cortical thick ascending limb of the loop of Henle", 
                  "distal convoluted tubule", 
                  "the short descending limb of the loop of Henle", 
                  "long descending limb of the loop of Henle in the outer medulla", 
                  "long descending limb of the loop of Henle in the inner medulla", 
                  "inner medullary collecting duct", 
                  "medullary thick ascending limb of the loop of Henle", 
                  "outer medullary collecting duct", 
                  "the initial segment of the proximal tubule", 
                  "proximal straight tubule in cortical medullary rays", 
                  "last segment of the proximal straight tubule in the outer stripe of medulla")
)

tubule_region_info
```

```{r}
knitr::include_graphics("../data/kidney/kidney-map/mouse-tubule/tubule_region.png")
```



```{r}
# quality check
tubule_pred$scores %>% 
    pheatmap::pheatmap(angle_col = 315, 
                       cluster_cols = T, 
                       color = viridis::viridis(n = 1000, option = "D"))

plotDeltaDistribution(tubule_pred, 
                      show = "delta.med")

marker_low_assignments <- pruneScores(results = tubule_pred, 
                                      nmads = 3) %>% table
```


```{r}
colData(subs_dev_cell_tubule)$location <- tubule_pred$labels
colData(subs_dev_cell_tubule)$assignment_quality_score <- ifelse(pruneScores(tubule_pred) == FALSE, "high", "low")
colData(subs_dev_cell_tubule)
```

```{r}
# assign the location information back to the dev_cell_sce object
others <- dev_cell_sce[, !colData(dev_cell_sce)$cell_type %in% cell_to_annotate] 
others %>% colData() %>% colnames(); subs_dev_cell_tubule %>% colData %>% colnames()
colData(others)$location <- ""
colData(others)$assignment_quality_score <- ""
dev_cell_sce2 <- cbind(others, subs_dev_cell_tubule)

# remove low assignment score cells
dev_cell_sce2 %>% colData() %>% .$assignment_quality_score %>% table(useNA = "ifany")
dev_cell_sce2 <- dev_cell_sce2[, which(colData(dev_cell_sce2)$assignment_quality_score != "low")]


# assese the assignment quality 
table(colData(subs_dev_cell_tubule[, subs_dev_cell_tubule$assignment_quality_score == "high"])$region, 
      colData(subs_dev_cell_tubule[, subs_dev_cell_tubule$assignment_quality_score == "high"])$location) %>% 
    prop.table(margin = 2) %>% round(digits = 3)

# annotate cells with location
colData(dev_cell_sce2)$location_cell <- ifelse(nchar(colData(dev_cell_sce2)$location) == 0, 
                                               colData(dev_cell_sce2)$cell_type, 
                                               paste(colData(dev_cell_sce2)$location, colData(dev_cell_sce2)$cell_type, sep = "_"))
dev_cell_sce2 %>% colData() 

library(scater)
plotReducedDim(dev_cell_sce2, 
               dimred = "TSNE", 
               colour_by = "region", 
               text_by = "location_cell", 
               text_size = 2, 
               point_size = 0.5)

saveRDS(dev_cell_sce2, "../data/kidney/kidney-map/dev-cell/dev_cell_sce.rds")
```


# prepare glomerular cells
```{r}
# GSE146912

library(magrittr)
library(dplyr)
library(data.table)
library(tibble)
library(SingleCellExperiment)

# normal
normal <- fread("../data/kidney/kidney-map/GSE146912/Normal_1.txt.gz", 
                sep = "\t", 
                header = TRUE)
gene_df <- normal[, 1:2]
gene_df <- gene_df %>% dplyr::rename(ensembl = IGIS) 
rownames(gene_df) <- gene_df$ensembl
cell_df <- data.frame(cell = colnames(normal)[-c(1,2)])
rownames(cell_df) <- cell_df$cell
normal %>% dim
normal[1:10, 1:5]
normal <- normal %>% column_to_rownames("IGIS") %>% select(-SYMBOL)

glomer_sce <- SingleCellExperiment(assays = list(counts = as(as.matrix(normal), "CsparseMatrix")), 
                                   rowData = gene_df, 
                                   colData = cell_df)
glomer_sce
saveRDS(glomer_sce, "../data/kidney/kidney-map/GSE146912/glomer_sce.rds")
```

```{r}
glomer_sce <- readRDS("../data/kidney/kidney-map/GSE146912/glomer_sce.rds")

# quality control
library(scran)
library(BiocParallel)
library(magrittr)
library(scDblFinder)
library(scater)
library(bluster)

is_mito <- grepl("^mt-", rowData(glomer_sce)$SYMBOL, ignore.case = F)
is_mito %>% table(useNA = "ifany")
rowData(glomer_sce)$SYMBOL[is_mito] %>% sort
colData(glomer_sce)$percent_mito <- round((colSums(assay(glomer_sce[is_mito, ], 1)) / colSums(assay(glomer_sce, 1))) * 100, digits = 3)
colData(glomer_sce)

qclusters <- quickCluster(glomer_sce, 
                          min.size = 100, 
                          method = "igraph", 
                          use.ranks = T, 
                          BPPARAM = MulticoreParam(workers = 12))
qclusters %>% table
colData(glomer_sce)$qclusters <- qclusters

qc <- perCellQCMetrics(glomer_sce, subsets = list(mito = is_mito))
qc$subsets_mito_percent %>% quantile(probs = seq(0, 1, 0.05))

qc_stat <- perCellQCFilters(x = qc, 
                            sub.fields = "subsets_mito_percent")
qc_stat %>% as.data.frame() %>% colSums()
glomer_sce <- glomer_sce[, !(qc_stat$low_n_features | qc_stat$low_lib_size | qc$subsets_mito_percent > 5)]
```

```{r}
# doublet detection
dbs <- scDblFinder(glomer_sce)
dbs$scDblFinder.class %>% table
glomer_sce <- glomer_sce[, dbs$scDblFinder.class == "singlet"]
glomer_sce
```


```{r}
# normalization
libsize <- calculateSumFactors(assay(glomer_sce, 1), clusters = glomer_sce$qclusters)
libsize %>% range
sizeFactors(glomer_sce) <- libsize

glomer_sce <- logNormCounts(glomer_sce)
glomer_sce
```


```{r}
# model gene variance
glomer_variance2 <- modelGeneVar(glomer_sce)
glomer_variance2[order(glomer_variance2$bio, decreasing = T), ]

# feature selection
hvgs <- getTopHVGs(glomer_variance2, fdr.threshold = 0.01)
hvgs %>% str
```


```{r}
# dimensionality reduction
pcs <- calculatePCA(assay(glomer_sce, 2), 
                    subset_row = hvgs)
percentvar <- attr(pcs, "percentVar")
library(PCAtools)
elbow <- findElbowPoint(percentvar)

df <- data.frame(pc = 1:length(percentvar), 
                 percentvar = percentvar)

df %>% 
    ggplot(aes(pc, percentvar)) +
    geom_line() +
    geom_point() +
    geom_vline(xintercept = elbow, linetype = 2)

dpcs <- getDenoisedPCs(x = assay(glomer_sce, 2), 
                       technical = glomer_variance2, 
                       subset.row = hvgs, 
                       BPPARAM = MulticoreParam(12))
dpcs %>% names
dpcs$components %>% dim
percentvar <- dpcs$percent.var
elbow <- findElbowPoint(percentvar)
data.frame(PC = 1:length(percentvar), 
           percentvar = percentvar) %>% 
    ggplot(aes(PC, percentvar)) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept = elbow, linetype = 2)

reducedDim(glomer_sce, "PCA") <- dpcs$components[, 1:elbow]
glomer_sce <- runTSNE(glomer_sce, use.dimred = "PCA", perplexity = 10)

plotReducedDim(glomer_sce, 
               dimred = "TSNE", 
               colour_by = "qclusters", 
               point_size = 0.5)
```


```{r}
# clustering
clusters <- clusterRows(reducedDim(glomer_sce, "TSNE"), 
                        BLUSPARAM = NNGraphParam(cluster.fun = "louvain",
                                                 k = 50, 
                                                 BPPARAM = MulticoreParam(10)))
clusters %>% table

colLabels(glomer_sce) <- factor(clusters)
```


```{r}
source("R/idconv.R")
library(patchwork)
plotReducedDim(glomer_sce, 
               dimred = "TSNE", 
               # colour_by = "percent_mito",
               colour_by = idconv("Adrb1"),
               text_by = "label", 
               point_size = 0.4)  
plotReducedDim(glomer_sce, 
               dimred = "TSNE", 
               colour_by = idconv("Ehd3"), 
               text_by = "label", 
               point_size = 0.4)

plotExpression(glomer_sce, 
               features = idconv("Pax8"), 
               x = "label",
               colour_by = "label")
```

```{r}
# cell type annotation
cell_type <- c("1" = "Mesangial", 
               "2" = "parietal Endothelia", 
               "3" = "arteriolar Endothelia",
               "4" = "Mesangial",
               "5" = "Podocyte", 
               "6" = "PEC",
               "7" = "Podocyte", 
               "8" = "Podocyte", 
               "9" = "capillary Endothelia", 
               "10" = "capillary Endothelia", 
               "11" = "Immune cells",
               "12" = "capillary Endothelia", 
               "13" = "capillary Endothelia", 
               "14" = "VSMC")

glomer_sce$celltype <- cell_type[glomer_sce$label]
glomer_sce %>% colData

# remove immune cells
glomer_sce <- glomer_sce[, !(glomer_sce$celltype == "Immune cells")]

saveRDS(glomer_sce, "../data/kidney/kidney-map/GSE146912/glomer_sce.rds")
```



```{r}
# glomerular 2021 nature communication
# GSE160048

unbiased <- read.table("../data/kidney/kidney-map/glomer2021/GSE160048_mouse_glom_single_cell_unbiased_rpkms.txt", 
                       sep = "\t", 
                       header = T)
unbiased <- unbiased %>% separate(col = X, into = c("symbol", "uniprot"), sep = "\\|")
unbiased <- unbiased[, -2]

# remove duplicated symbols
unbiased <- aggregate(unbiased[, -1], by = list(symbol = unbiased[, 1]), FUN = sum)
unbiased$symbol %>% duplicated() %>% table()

# ERCC spike-in matrix
ids <- grepl("^ERCC_", unbiased$symbol, ignore.case = F)
ids %>% table
ercc_matrix <- unbiased[ids, ]
unbiased[!ids, -1] %>% colSums() %>% quantile(probs = seq(0, 1, 0.1))

# convert rpkm to lognorm
gene_length <- readRDS("gene_length.rds")
gene_length[unbiased$symbol %in% gene_length$symbol, ]
gene_length$symbol %>% duplicated() %>% table

unbiased <- unbiased[!ids, ]
unbiased[1:10, 1:5] 
```
$$rpkm = \frac{reads \space counts}{\frac{sequencing \space deeps}{1e6}  * gene \space length \space (kb)}$$






# prepare nuron cells
```{r}
# from 
library(fs)
library(data.table)
library(magrittr)
library(dplyr)

glomer_unbiased <- read.table("../data/kidney/kidney-map/glomer2021/GSE160048_mouse_glom_single_cell_unbiased_rpkms.txt")
glomer_unbiased %>% dim
glomer_unbiased %>% head

SCG_files <- dir_ls("../data/kidney/kidney-map/sym-sen2022/", regexp = "_SCG_")
DRG_files <- dir_ls("../data/kidney/kidney-map/sym-sen2022/", regexp = "_DRG_")

cell_info <- read.csv("../data/kidney/kidney-map/sym-sen2022/GSE175421_cell-barcode-ids.csv")

library(EnsDb.Mmusculus.v79)
cell_info %>% head

scg3_cell_info <- cell_info %>% dplyr::filter(library.id == "SCG_3")
scg3_cell_info <- scg3_cell_info %>% 
    dplyr::rename("barcode" = X) %>% 
    mutate(barcode = sub("_[0-9]_[0-9]$", "", barcode))

drg3_cell_info <- cell_info %>% dplyr::filter(library.id == "DRG_3")
drg3_cell_info <- drg3_cell_info %>% 
    dplyr::rename("barcode" = X) %>% 
    mutate(barcode = sub("_[0-9]_[0-9]$", "", barcode))


scg3 <- fread(SCG_files[[3]], header = T) %>% as.data.frame()
scg3[1:10, 1:10]
sel_index <- (colnames(scg3[, -1]) %in% scg3_cell_info$barcode)
sel_index %>% head
scg_counts <- scg3[, -1] %>% .[, sel_index]
scg_counts$symbol <- scg3$GENE
scg_counts <- scg_counts %>% relocate(symbol)
scg_counts[1:10, 1:10]
scg_counts <- aggregate(scg_counts[, -1], by = list(symbol = scg_counts$symbol), FUN = sum)

rowdata_scg3 <- data.frame(symbol = scg_counts$symbol)
rowdata_scg3$ensembl <- mapIds(EnsDb.Mmusculus.v79, 
                               keys = rowdata_scg3$symbol, 
                               keytype = "SYMBOL", 
                               column = "GENEID")
rowdata_scg3 %>% head
scg_counts <- left_join(rowdata_scg3, scg_counts)
scg_counts <- scg_counts %>% dplyr::filter(!is.na(ensembl))
scg_counts$ensembl %>% duplicated() %>% table
rownames(scg_counts) <- scg_counts$ensembl
scg_counts <- scg_counts[, -(1:2)]

scg_sce <- SingleCellExperiment(assays = list(counts = as(as.matrix(scg_counts), "CsparseMatrix")), 
                                rowData = DataFrame(ensembl = rownames(scg_counts)), 
                                colData = scg3_cell_info %>% dplyr::filter(barcode %in% colnames(scg_counts)))

# filtering out nu-nuron cells
scg_sce %>% colData %>% .$main_cluster %>% table
cells_to_remove <- c("Fibroblasts", "Macrophages", "T Cells", "Vascular Endothelial Cells", "Sensory Neurons")
scg_sce <- scg_sce[, !(scg_sce$main_cluster %in% cells_to_remove)]
saveRDS(scg_sce, "../data/kidney/kidney-map/sym-sen2022/scg_sce.rds")


drg3 <- fread(DRG_files[[3]], header = T) %>% as.data.frame()
drg3[1:10, 1:10]
sel_index <- (colnames(drg3[, -1]) %in% drg3_cell_info$barcode)
drg_counts <- drg3[, -1] %>% .[, sel_index]
drg_counts$symbol <- drg3$GENE
drg_counts <- drg_counts %>% relocate(symbol)
drg_counts[1:10, 1:10]
drg_counts <- aggregate(drg_counts[, -1], by = list(symbol = drg_counts$symbol), FUN = sum)

rowdata_drg3 <- data.frame(symbol = drg_counts$symbol)
rowdata_drg3$ensembl <- mapIds(EnsDb.Mmusculus.v79, 
                               keys = rowdata_drg3$symbol, 
                               keytype = "SYMBOL", 
                               column = "GENEID")
drg_counts <- left_join(rowdata_drg3, drg_counts)
drg_counts <- drg_counts %>% dplyr::filter(!is.na(ensembl))
drg_counts$ensembl %>% duplicated() %>% table
rownames(drg_counts) <- drg_counts$ensembl
drg_counts <- drg_counts[, -(1:2)] 

drg_sce <- SingleCellExperiment(assays = list(counts = as(as.matrix(drg_counts), "CsparseMatrix")), 
                                rowData = DataFrame(ensembl = rownames(drg_counts)), 
                                colData = drg3_cell_info %>% dplyr::filter(barcode %in% colnames(drg_counts)))

# remove un-nuron cells
cells_to_remove <- c("Fibroblasts", "Macrophages", "Sympathetic Neurons", "T Cells", "Vascular Endothelial Cells")
drg_sce$main_cluster %>% table

drg_sce <- drg_sce[, !c(drg_sce$main_cluster %in% cells_to_remove)]
saveRDS(drg_sce, "../data/kidney/kidney-map/sym-sen2022/drg_sce.rds")
```








